transform/cli/cli.go

```diff
@@ -1,79 +1,118 @@
 package cli
 
 import (
+	"bufio"
 	"bytes"
 	"encoding/json"
 	"fmt"
-	"github.com/sirupsen/logrus"
 	"io"
 	"os"
 
+	"github.com/sirupsen/logrus"
+
 	"github.com/konveyor/crane-lib/transform"
 	"github.com/konveyor/crane-lib/transform/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+var (
+	stdErr io.Writer
+	stdOut io.Writer
+	reader io.Reader
+	exiter func(int)
+)
+
+func init() {
+	stdErr = os.Stderr
+	stdOut = os.Stdout
+	reader = os.Stdin
+
+	exiter = os.Exit
+}
+
 type CustomPlugin struct {
 	// TODO: figure out a way to include the name of the plugin in the error messages.
-	name    string
-	runFunc func(*unstructured.Unstructured) (transform.PluginResponse, error)
+	metadata transform.PluginMetadata
+	runFunc  func(*unstructured.Unstructured) (transform.PluginResponse, error)
 }
 
-func (c *CustomPlugin) Run(u *unstructured.Unstructured) (transform.PluginResponse, error) {
+func (c *CustomPlugin) Run(u *unstructured.Unstructured, extras map[string]string) (transform.PluginResponse, error) {
 	if c.runFunc == nil {
 		return transform.PluginResponse{}, nil
 	}
 	return c.runFunc(u)
 }
 
-func (c *CustomPlugin) Name() (string) {
-	return c.name
+func (c *CustomPlugin) Metadata() transform.PluginMetadata {
+	return c.metadata
 }
-func NewCustomPlugin(name string, runFunc func(*unstructured.Unstructured) (transform.PluginResponse, error)) transform.Plugin {
+
+func NewCustomPlugin(name, version string, optionalFields []string, runFunc func(*unstructured.Unstructured) (transform.PluginResponse, error)) transform.Plugin {
 	return &CustomPlugin{
-		name:    name,
+		metadata: transform.PluginMetadata{
+			Name:            name,
+			Version:         version,
+			RequestVersion:  []transform.Version{transform.V1},
+			ResponseVersion: []transform.Version{transform.V1},
+			OptionalFields:  optionalFields,
+		},
 		runFunc: runFunc,
 	}
 }
 
-func Unstructured(reader io.Reader) (*unstructured.Unstructured, error) {
-	decoder := json.NewDecoder(reader)
-	u := &unstructured.Unstructured{}
-	err := decoder.Decode(u)
-	if err != nil {
-		return nil, &errors.PluginError{
-			Type:         errors.PluginInvalidIOError,
-			Message:      "unable to decode valid json from the reader",
-			ErrorMessage: err.Error(),
-		}
-	}
-	return u, nil
+// Will write the error the standard error and will exit with 1
+func WriterErrorAndExit(err error) {
+	fmt.Fprint(stdErr, err.Error())
+	exiter(1)
 }
 
-func ObjectReaderOrDie() io.Reader {
-	return os.Stdin
+func Logger() logrus.FieldLogger {
+	return &logrus.Logger{}
 }
 
-func stdOut() io.Writer {
-	return os.Stdout
-}
+func RunAndExit(plugin transform.Plugin) {
+	// Get the reader from Standard In.
+	//var s string
+	scan := bufio.NewScanner(reader)
+	ok := scan.Scan()
```

Comment:
this will break plugins that take json input over multiple lines. for example when I try the openshift plugin in the example, the first try fails, second works:

```
Alays-MacBook-Pro-2:openshift alay$ ./openshift 
{
{"type":"PluginInvalidInputError","message":"error writing plugin response to stdOut","error":"unexpected EOF"}Alays-MacBook-Pro-2:openshift alay$ ./openshift 
{"apiVersion": "v1"}
{"type":"PluginInvalidInputError","message":"error writing plugin response to stdOut","error":"Object 'Kind' is missing in '{\"apiVersion\": \"v1\"}'"}Alays-MacBook-Pro-2:openshift alay$ 
```